packages/eslint-plugin/tests/rules/no-invalid-this.test.ts

```diff
@@ -0,0 +1,465 @@
+import { AST_NODE_TYPES } from '@typescript-eslint/experimental-utils';
+import rule from '../../src/rules/no-invalid-this';
+import { RuleTester, getFixturesRootDir } from '../RuleTester';
+
+const rootPath = getFixturesRootDir();
+
+const ruleTester = new RuleTester({
+  parser: '@typescript-eslint/parser',
+  parserOptions: {
+    sourceType: 'module',
+    tsconfigRootDir: rootPath,
+    project: './tsconfig.json',
+  },
+});
+
+const error = {
+  messageId: 'unexpectedThis',
+};
```

Comment:
both of your errors are probably because this is typed as `{ messageId: string }`.
`string` is not assignable to the more constrainted type `'unexpectedThis'`.

```suggestion
const error = {
  messageId: 'unexpectedThis' as const,
};
```

Reply:
ok, but how this is different from other rules. I checked other extended rules and also this base rule couldn't find any difference which will lead this. I am surely missing something

Reply:
You have misread the code.
The ones that pre-declare an error object do so purely because the base rule doesn't use messageId. Additionally, they type the object as `any`.

In those rules, this is done because our types are designed to enforce the best-practice of using `messageId` always. So because those rules don't have a messageId (yet,  they will with the ESLint v7 release), we have to work around it with an `any`.

Example:
https://github.com/typescript-eslint/typescript-eslint/blob/188b68979d114ba0957a076fe9c27d154223512c/packages/eslint-plugin/tests/rules/quotes.test.ts#L22-L34

I would _strongly_ recommend **not** pre-declaring an error object, and just declaring the object inline with the tests, as you would normally.

Reply:
> The ones that pre-declare an error object do so purely because the base rule doesn't use messageId. Additionally, they type the object as `any`.

ohh, ook, cool I will change it